pennylane/compiler/compiler.py

```diff
@@ -0,0 +1,134 @@
+
+
+def available_compilers() -> List[str]:
+    """Loads and returns a list of available compilers that are
+    installed and compatible with the :func:`~.qjit` decorator.
+
+    **Example**
+
+    This method returns the name of installed compiler packages supported in
+    PennyLane. For example, after installing the `Catalyst <https://github.com/pennylaneai/catalyst>`__
+    compiler, this will now appear as an available compiler:
+
+    >>> qml.compiler.available_compilers()
+    ['catalyst']
+    """
+
+    # This class updates the class variable names_entrypoints
+    _refresh_compilers()
+
+    return list(AvailableCompilers.names_entrypoints.keys())
+
+
+def available(compiler="catalyst") -> bool:
+    """Check the availability of the given compiler package.
+
+    Args:
+        compiler (str): name of the compiler package (default value is ``catalyst``)
+
+    Return:
+        bool: ``True`` if the compiler package is installed on the system
+
+    **Example**
+
+    Before installing the ``pennylane-catalyst`` package:
+
+    >>> qml.compiler.available("catalyst")
+    False
+
+    After installing the ``pennylane-catalyst`` package:
+
+    >>> qml.compiler.available("catalyst")
+    True
+    """
+
+    # It only refreshes the compilers names and entry points if the name
+    # is not already stored. This reduces the number of re-importing
+    # ``pkg_resources`` as it can be a very slow operation on systems
+    # with a large number of installed packages.
+
+    if compiler not in AvailableCompilers.names_entrypoints:
+        # This class updates the class variable names_entrypoints
+        _refresh_compilers()
+
+    return compiler in AvailableCompilers.names_entrypoints
+
+
+def active(compiler="catalyst") -> bool:
+    """Check whether the caller is inside a :func:`~.qjit` evaluation context.
+
+    This helper function may be used during implementation
+    to allow differing logic for circuits or operations that are
+    just-in-time compiled versus those that are not.
+
+    Args:
+        compiler (str): name of the compiler package (default value is ``catalyst``)
+
+    Return:
+        bool: True if the caller is inside a QJIT evaluation context
+
+    **Example**
+
+    In the JIT compilation of PennyLane programs, this helper method checks
+    the status of the compilation. For example, Catalyst captures Python
+    programs by tracing and we can tell if the caller is inside the Catalyst
+    tracer context manager or not but calling this method.
+
+    This method is practically useful in implementing quantum operations to
+    correctly call the interpreter or compiler equivalent functions.
+    """
+
+    compilers = AvailableCompilers.names_entrypoints
+
+    if not compilers:
+        return False
+
+    try:
+        tracer_loader = compilers[compiler]["context"].load()
+        return tracer_loader.is_tracing()
+    except KeyError:
+        return False
```

Comment:
Would a key error here not indicate that a broken or incompatible plugin is loaded here? I'm wondering if we actually do want to raise such an issue.

Reply:
This is removed in PR #4698 by checking the entry-points interface in `_refresh_compilers`.

Reply:
Ah this complicates the review process if we just have outdated code here that is updated elsewhere anyways. Would you be able to move the relevant changes from the other PRs here?